tracer/src/Datadog.Trace/Debugger/Instrumentation/Collections/ProbeDataCollection.cs

```diff
@@ -42,52 +42,78 @@ internal static ProbeMetadataCollection Instance
         }
 
         /// <summary>
-        /// Tries to create a new <see cref="ProbeData"/> at <paramref name="index"/>.
+        /// Tries to create a new <see cref="ProbeData"/> at <paramref name="index"/> and return it.
         /// </summary>
         /// <param name="index">The index of the probe inside <see cref="EverGrowingCollection{TPayload}.Items"/></param>
         /// <param name="probeId">The id of the probe/></param>
         /// <returns>true if succeeded (either existed before or just created), false if fails to create</returns>
-        public bool TryCreateProbeMetadataIfNotExists(int index, string probeId)
+        public ref ProbeData TryCreateProbeDataIfNotExists(int index, string probeId)
         {
-            if (IsIndexExists(index))
+            ref var probeData = ref TryGetProbeDataIndex(index, probeId);
+
+            if (!probeData.IsEmpty())
             {
-                return true;
+                return ref probeData;
             }
 
             // Create a new one at the given index
             lock (ItemsLocker)
             {
-                if (IsIndexExists(index))
+                probeData = ref TryGetProbeDataIndex(index, probeId);
+
+                if (!probeData.IsEmpty())
                 {
-                    return true;
+                    return ref probeData;
                 }
 
                 EnlargeCapacity(index);
 
                 if (Log.IsEnabled(Vendors.Serilog.Events.LogEventLevel.Debug))
                 {
-                    Log.Debug($"{nameof(ProbeMetadataCollection)}.{nameof(TryCreateProbeMetadataIfNotExists)}: Creating a new probe metadata info for index = {index}, Items.Length = {Items.Length}");
+                    Log.Debug($"{nameof(ProbeDataCollection)}.{nameof(TryCreateProbeDataIfNotExists)}: Creating a new probe metadata info for index = {index}, Items.Length = {Items.Length}");
```

Comment:
Now we've merged the Serilog analyzers in #3781 and #3785, this log will break the build. You musn't use interpolation (it's uglier, I know, but unfortunately necessary)
```suggestion
                    Log.Debug(nameof(ProbeDataCollection) + "." + nameof(TryCreateProbeDataIfNotExists) + ": Creating a new probe metadata info for index = {Index}, Items.Length = {Length}", index, Items.Length);
```

Reply:
Done with rebase.